### PR TITLE
logger: disable sof-logger for zephyr

### PR DESCRIPTION
**Pull request by keqiaozhang**
since zephyr doesnot support sof-logger, so add a function
to disable logger trace for zephyr during the test.

Signed-off-by: Zhang Keqiao <keqiao.zhang@intel.com>

### CODE REVIEW

File: case-lib/lib.sh [hunk 1]
```diff
@@ -343,3 +343,19 @@ disable_kernel_check_point()
 {
     KERNEL_CHECKPOINT="disabled"
 }
+
+is_sof_logger_disabled()
```
Comment (marc-hb) [inline case-lib/lib.sh:347]: I think `logger_disabled()` is long enough name.

File: case-lib/lib.sh [hunk 2]
```diff
@@ -343,3 +343,19 @@ disable_kernel_check_point()
...
+{
+    # get the firmware type
+    manifest=/etc/sof/manifest.txt
+    FIRMWARE_TYPE=$(grep "firmwaretype" < "$manifest" | awk -F ':' '{print $2}' | sed -e 's/^[ ]*//g;s/\"//g')
+    # disable sof-logger for Zephyr built firmware
+    if [ "$FIRMWARE_TYPE" == "zephyr" ]; then
```
Comment (marc-hb) [inline case-lib/lib.sh:353]: This  above seems a bit long and over complicated, could this be shorten into something like this:
```shell
if grep `firmwaretype.*zephyr` /etc/sof/manifest.txt; then
```
Just a really simplified example: you may want a slightly more complex regular expression.
```

File: case-lib/lib.sh [hunk 3]
```diff
@@ -343,3 +343,19 @@ disable_kernel_check_point()
 {
     KERNEL_CHECKPOINT="disabled"
 }
+
+is_sof_logger_disabled()
+{
+    # get the firmware type
+    manifest=/etc/sof/manifest.txt
+    FIRMWARE_TYPE=$(grep "firmwaretype" < "$manifest" | awk -F ':' '{print $2}' | sed -e 's/^[ ]*//g;s/\"//g')
```
Comment (marc-hb) [inline case-lib/lib.sh:351]: No need for the `<`.

No need for the `grep`, `awk` can already do `grep` (and many other things): `awk -F ':' '/firmwaretype/ {print $2 }'`

Regular expressions with only special characters are time-consuming to read and deserve a comment, for instance `# remove whitespace`.

**Review by marc-hb — CHANGES_REQUESTED** [review]
Can you please add one more `is_zephyr()` function? I mean `logger_disabled()` invokes `is_zephyr()` like this:
```shell
if is_zephyr; then
   return 0
else ...
```

Can you please give an example of this `/etc/sof/manifest.txt` file and a link to what code creates and deploys it?

This `/etc/sof/manifest.txt` file may be good enough for CI in the very short term but I don't think it will work for most developers. I compile with and without Zephyr many times every day and I also use `sof-test` yet my `/etc/sof/` has been empty since forever.


@andyross , @kv2019i , @lyakh : any other, simple, automatable and reliable way to tell the difference between plain SOF versus SOF+Zephyr?

PS: I've been using a `.tarball-version` hack that shows up in dmesg but this won't work for CI

File: case-lib/lib.sh (hunk 1 above)
Comment (keqiaozhang) [reply case-lib/lib.sh:347]: @marc-hb ， /etc/sof/manifest.txt has below information:
```
{
    "date" : "2021-05-21 07:35:42 UTC",
    "version" : {
        "kernelBranch" : "topic/sof-dev",
        "kernelCommit" : "dcea8052",
        "sofBranch" : "main",
        "sofCommit" : "72a51a462fb8"
        "firmwaretype" : "zephyr"
    },
    "filepath" : {
        "kernel" : "http://sof-ci-storage.sh.intel.com/release/linux/topic/sof-dev/dcea8052/linux-image-5.13.0-rc1-daily-nocodec-20210520-0_5.13.0-rc1-topic-sof-dev-dcea8052_amd64.deb",
        "firmware" : "http://sof-ci-storage.sh.intel.com/release/sof/main/72a51a462fb8/fw/zephyr/sof-apl.ri",
        "ldc" : "http://sof-ci-storage.sh.intel.com/release/sof/main/72a51a462fb8/fw/zephyr/sof-apl.ldc",
        "lst" : "http://sof-ci-storage.sh.intel.com/release/sof/main/72a51a462fb8/fw/zephyr/sof-apl.lst",
        "topology" : "http://sof-ci-storage.sh.intel.com/release/sof/main/72a51a462fb8/tplg/sof-apl-nocodec.tplg",
        "logger" : "http://sof-ci-storage.sh.intel.com/release/sof/main/72a51a462fb8/tools/sof-logger",
        "ctl" : "http://sof-ci-storage.sh.intel.com/release/sof/main/72a51a462fb8/tools/sof-ctl"
    }
}
```
The firmwaretype is added in the deployment stage. related code is in sof-sh/deploy.sh:
```
# generate manifest file
save_manifest() {
    # save Manifest file to DUT
    cat << END > $MANIFEST_FILE
{
    "date" : "$TODAY",
    "version" : {
        "kernelBranch" : "$KERNEL_BRANCH",
        "kernelCommit" : "$LINUX_SHA",
        "sofBranch" : "$SOF_BRANCH",
        "sofCommit" : "$SOF_SHA"
        "firmwaretype" : "$FW_TYPE"
    },
    "filepath" : {
        "kernel" : "$KERNEL_URL",
        "firmware" : "$FW_URL",
        "ldc" : "$LDC_URL",
        "lst" : "$LST_URL",
        "topology" : "$TPLG_URL",
        "logger" : "$LOGGER_URL",
        "ctl" : "$CTL_URL"
    }
}
END
```
so we will generate the manifest.txt every time when deploy sof environment.

File: case-lib/lib.sh (hunk 2 above)
Comment (keqiaozhang) [reply case-lib/lib.sh:353]: Good point, I will refine my patch.

Comment (marc-hb) [reply case-lib/lib.sh:353]: I didn't know `manifest.txt` was actually a `manifest.json`. The standard, command-line parser for JSON is `jq`, please drop awk, grep etc. and use `jq`:

```
apt-get install jq
jq '.version.firmwaretype' manifest.txt
 => "zephyr"
```

File: case-lib/lib.sh (hunk 1 above)
Comment (aiChaoSONG) [reply case-lib/lib.sh:347]: @keqiaozhang Please align variable naming style, `fwType` or `firmwareType`

Comment (keqiaozhang) [reply case-lib/lib.sh]: Updated.

File: case-lib/lib.sh (hunk 2 above)
Comment (keqiaozhang) [reply case-lib/lib.sh]: Updated. Thanks!

File: case-lib/lib.sh (hunk 1 above)
Comment (kv2019i) [reply case-lib/lib.sh]: @keqiaozhang To facilitate running sof-test standalone, the manifest file should be documented somewhere within sof-test repository (an example manifest file perhaps).

File: case-lib/lib.sh [hunk 4]
```diff
@@ -343,3 +343,26 @@ disable_kernel_check_point()
...
 }
+
+is_zephyr()
+{
+    # check if jq is installed, will remove this part
+    # after all DUTs have jq installed.
+    hash jq 2>/dev/null || sudo apt install jq -y &>/dev/null
+
+    manifest=/etc/sof/manifest.txt
+    jq '.version.firmwareType' "$manifest" | grep "zephyr"
```
Comment (lyakh) [inline case-lib/lib.sh:354]: how about adding the OS (XTOS / Zephyr) information to firmware extended manifest?

**Review by kv2019i — COMMENTED** [review]
I tested the patch on a system with Zephyr firmware and it did work as expected. We should make the manifest-based configuration a bit cleaner so it makes sense for standalone usage of sof-test, but otherwise this looks good.

File: case-lib/lib.sh [hunk 5]
```diff
@@ -343,3 +343,26 @@ disable_kernel_check_point()
 {
     KERNEL_CHECKPOINT="disabled"
 }
+
+is_zephyr()
+{
+    # check if jq is installed, will remove this part
+    # after all DUTs have jq installed.
+    hash jq 2>/dev/null || sudo apt install jq -y &>/dev/null
```
Comment (marc-hb) [inline case-lib/lib.sh:351]: Nit: with `type -p` instead of `hash` you can redirect only stdout and not redirect stderr which is "safer".

There's no need to redirect the output of `apt install`, it's not very verbose, it's good to see when it happened and it's even more important to see any error like proxy configuration or network glitch or dependency problem - many things could go wrong. This is test code, the output does not have to be very "tidy", seeing what's going on is much more important.

File: case-lib/lib.sh [hunk 6]
```diff
@@ -343,3 +343,26 @@ disable_kernel_check_point()
...
+
+    manifest=/etc/sof/manifest.txt
```
Comment (marc-hb) [inline case-lib/lib.sh:353]: make the variable `local` to avoid clashes.

File: case-lib/lib.sh [hunk 7]
```diff
@@ -343,3 +343,26 @@ disable_kernel_check_point()
...
+{
+    # disable logger for zephyr
+    if is_zephyr; then
+        return 0
+    # disable sof-logger when -s is set to 0
+    elif [[ ${OPT_VAL['s']} -ne 1 ]]; then
+        return 0
+    else
+        return 1
+    fi
```
Comment (marc-hb) [inline case-lib/lib.sh:367]: Style nit/personal preference:
```shell
if is_zephyr || [[ ${OPT_VAL['s']} -ne 1 ]]; then
   return 0
fi

return 1

```

File: case-lib/lib.sh (hunk 4 above)
Comment (marc-hb) [reply case-lib/lib.sh:354]: `grep -q`.

I think there's also a `test -e "$manifest" || return 1` missing before the `jq`.

**Review by marc-hb — CHANGES_REQUESTED** [review]
I suggest we merge this with the manifest.txt first to get CI started and then try to find a better test later. Only the `is_zephyr()` function will have to change.

File: case-lib/lib.sh (hunk 7 above)
Comment (keqiaozhang) [reply case-lib/lib.sh:365]: Concise and reasonable. Updated

File: case-lib/lib.sh (hunk 4 above)
Comment (keqiaozhang) [reply case-lib/lib.sh:355]: > how about adding the OS (XTOS / Zephyr) information to firmware extended manifest?

I'm OK with that, if we all agree, I can add it in deploy.sh.

File: case-lib/lib.sh (hunk 1 above)
Comment (keqiaozhang) [reply case-lib/lib.sh]: @kv2019i , most of our test scripts have "-s" option, you can set it as 0 to skip the trace collection. for example:
```
./check-playback.sh -d 10 -l 1 -r 1 -s 0
```
if you want to run "run-all-tests.sh" without logger, we can figure out a new solution for it next.

**Comment by aiChaoSONG** [comment]
> I think there's also a test -e "$manifest" || return 1 missing before the jq.

@marc-hb @keqiaozhang  For CI devices, $manifest must exist, otherwise, the deployment failed, so the test is not started.
For personal device, this file may not exist. so I think we should regard '$manifest not exist' as 'zephyr not exist', and continue to check other options. this is for CI and personal device compatibility

File: case-lib/lib.sh (hunk 1 above)
Comment (marc-hb) [reply case-lib/lib.sh]: >  If you want to run "run-all-tests.sh" without logger, we can figure out a new solution for it next.

I have a hack that does this here: https://github.com/marc-hb/sof-test/commit/run-all-nologger

**Review by marc-hb — APPROVED** [review]
LGTM but probably not for some others yet.

> For personal device, this file may not exist. so I think we should regard '$manifest not exist' as 'zephyr not exist', 

I think the majority of people who reviewed this PR use Zephyr without a manifest.txt and at least one had a good idea that he will submit immediately after this PR is merged :-)

> Edit: If we just want to test if the FW is zephyr, a grep is enough, we can eliminate the dependency on jq

jq is awesome, it's a great dependency and in this case it makes the manifest-based test very reliable. Now if the manifest.txt test goes away soon then I agree the `jq` dependency will be very short lived in this case...

**Review by kv2019i — APPROVED** [review]
Fair enough @marc-hb , let's go with this.

**Comment by keyonjie** [comment]
why does zephyr not support sof-logger? without FW logs, how can we debug the SOF issues?

**Comment by marc-hb** [comment]
We could make sof-logger compatible with Zephyr but the plan is to migrate to https://docs.zephyrproject.org/latest/reference/logging/

For now logging is provided by one of these: https://github.com/zephyrproject-rtos/zephyr/pull/35186

**Review by fredoh9 — APPROVED** [review]
didn't know sof-logger is not supported for Zephyr

**Comment by marc-hb** [comment]
3 approvals, 8 days in review, simple and easy to maintain change: merging.

Some very good suggestions and objections: please submit incremental changes.

File: case-lib/lib.sh (hunk 7 above)
Comment (marc-hb) [reply case-lib/lib.sh:365]: Actually, this can probably be just:
```shell
logger_disabled()
{
   is_zephyr || [[ ${OPT_VAL['s']} -ne 1 ]
}
```

**Comment by marc-hb** [comment]
SKIP check-sof-logger if is_zephyr #724

**Comment by marc-hb** [comment]
#809 just enabled sof-logger with Zephyr. Seems to work fine.
